src/goto-programs/string_refine_preprocess.cpp

```diff
@@ -0,0 +1,1343 @@
+/*******************************************************************\
+
+Module: Preprocess a goto-programs so that calls to the java String
+        library are recognized by the string solver
+
+Author: Romain Brenguier
+
+Date:   September 2016
+
+\*******************************************************************/
+
+#include "string_refine_preprocess.h"
+
+#include <util/std_expr.h>
+#include <util/symbol_table.h>
+#include <util/pointer_offset_size.h>
+#include <util/prefix.h>
+#include <goto-programs/class_identifier.h>
+#include <solvers/refinement/string_expr.h>
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::new_tmp_symbol
+
+  Inputs: a name and a type
+
+ Outputs: a new symbol
+
+ Purpose: add a temporary symbol with the given name and type to the
+          symbol table and returns it
+
+\*******************************************************************/
+
+symbol_exprt string_refine_preprocesst::new_tmp_symbol(
+  const std::string &prefix, const typet &type)
+{
+  auxiliary_symbolt tmp_symbol;
+  std::ostringstream buf;
+  buf << "string_preprocess_tmp" << (next_symbol_id++) << "#" << prefix;
+  std::string name=buf.str();
+  tmp_symbol.base_name=name;
+  tmp_symbol.is_static_lifetime=false;
+  tmp_symbol.mode=ID_java;
+  tmp_symbol.name=name;
+  tmp_symbol.type=type;
+  symbol_table.add(tmp_symbol);
+  return symbol_exprt(name, type);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::new_symbol
+
+  Inputs: a name and a type
+
+ Outputs: a new symbol
+
+ Purpose: add a symbol with the given name and type to the symbol table
+          and returns it
+
+\*******************************************************************/
+
+symbol_exprt string_refine_preprocesst::new_symbol(
+  const std::string &prefix, const typet &type, bool static_lifetime)
+{
+  symbolt tmp_symbol;
+  std::ostringstream buf;
+  buf << "string_preprocess" << (next_symbol_id++) << "#" << prefix;
+  std::string name=buf.str();
+  tmp_symbol.base_name=name;
+  tmp_symbol.is_static_lifetime=static_lifetime;
+  tmp_symbol.is_state_var=static_lifetime;
+  tmp_symbol.is_type=false;
+  tmp_symbol.is_lvalue=true;
+  tmp_symbol.mode=ID_java;
+  tmp_symbol.name=name;
+  tmp_symbol.type=type;
+  symbol_table.add(tmp_symbol);
+  return symbol_exprt(name, type);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::declare_function
+
+  Inputs: a name and a type
+
+ Purpose: declare a function with the given name and type
+
+\*******************************************************************/
+
+void string_refine_preprocesst::declare_function(
+  irep_idt function_name, const typet &type)
+{
+  auxiliary_symbolt func_symbol;
+  func_symbol.base_name=function_name;
+  func_symbol.is_static_lifetime=false;
+  func_symbol.mode=ID_java;
```

Comment:
Invocation of the preprocessing is not part of this pull request but of this one: https://github.com/diffblue/cbmc/pull/429 (in particular here: https://github.com/diffblue/cbmc/pull/429/files#diff-41db988013f70caad45ce489749c7df6R921)
This is because the original pull request was too big and thus was split into several parts.